casual/casual-api/src/main/java/se/laz/casual/api/queue/DequeueReturn.java

```
@@ -0,0 +1,123 @@
+package se.laz.casual.api.queue;
+
+import se.laz.casual.api.CasualRuntimeException;
+import se.laz.casual.api.flags.ErrorState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class DequeueReturn
+{
+    private final List<QueueMessage> queueMessages;
+    private final ErrorState errorState;
+
+    private DequeueReturn(List<QueueMessage> queueMessages, ErrorState errorState)
+    {
+        Objects.requireNonNull(queueMessages, "queueMessages can't be null");
+        Objects.requireNonNull(errorState, "errorState can't be null");
+        this.queueMessages = queueMessages;
+        this.errorState = errorState;
+    }
+
+    public List<QueueMessage> getQueueMessages()
+    {
+        if (ErrorState.OK.equals(errorState))
+        {
+            return queueMessages;
+        }
+        else
+        {
+            throw new CasualRuntimeException("Attempt to get queuemessages from queuereturn, but state was not OK");
```

Add the actual error in the message

---

Not sure we should go with a getter that can throw.
From a usage point of view, I'd not except that that can happen - an empty list, fine and if so then it is up to me, the user, to check the error state.

---

I did that to possibly prevent hiding errors, but I should probably make it return an empty list and leave it to the user to check ErrorState or not. In a way returning an empty list would be more similar to EnqueueReturn that returns Optional.empty(), but at the same time an empty list does not convey that something is amiss. I'll probably go with empty list and add some javadoc to clarify usage for any curious user

---

But if you retrieve the messages and there are none, from a user perspective not having it raise the error is a bit worrying to me. I prefer this initial implementation. 
Question - can you have messages and an error state? If yes, then move the error checking out - if not, then I would say leave it in.

---

I think that it is not exceptional to ask for messages from a queue, there are no messages, you get an empty list back.

---

Then again, that will also be covered by the ErrorState being OK

---

Still feels weird with a getter that throws but at least we still cover the case where there are no messages, that was the one I was considering

---

If the call to dequeue resulted in an error (currently that would be TPENOENT when the queue isn't accessible in any connection pool) there should never be any data to retrieve from the return wrapper. Throwing absolutely makes sense here. After some thought I'll probably keep the throwing, but add more info to the exception message. I'll also add some javadoc to the method to clarify expected behavior to users

Compared to the old behavior both enqueue and dequeue will require extra work for the end user. Based on my tests enqueue previously returned an all zero UUID if the queue did not exist and dequeue would return an empty list. While it's more work for the end user I do think it's better in the long run that errors are exposed and states with errors need to be handled explicitly